File: src/test/scala/com/exasol/cloudetl/kafka/ittests/KafkaTopicDataImporterAvroToJsonIT.scala
package com.exasol.cloudetl.kafka

import java.lang.{Integer => JInt}
import java.lang.{Long => JLong}

import com.exasol.ExaMetadata

import org.mockito.ArgumentMatchers
import org.mockito.ArgumentMatchers._
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.when
import org.mockito.invocation.InvocationOnMock
import org.mockito.stubbing.Answer

class KafkaTopicDataImporterAvroToJsonIT extends KafkaTopicDataImporterAvroIT {

  test("run emits records from starting initial offset") {
    val newProperties = properties ++ Map("AS_JSON_DOC" -> "true")
    createCustomTopic(topic)
    publishToKafka(topic, AvroRecord("{'Value':'abc'}", 3, 13))
    publishToKafka(topic, AvroRecord("{'Value':'hello'}", 4, 14))
    publishToKafka(topic, AvroRecord("{'Value':'xyz'}", 5, 15))

    val mockedIterator = mockExasolIterator(newProperties, Seq(0), Seq(-1))
    KafkaTopicDataImporter.run(getMockedMetadata(), mockedIterator)

    verify(mockedIterator, times(3)).emit(any(classOf[Array[Object]]))
    verify(mockedIterator, times(3)).emit(anyString(), anyInt().asInstanceOf[JInt], anyLong().asInstanceOf[JLong])

    verify(mockedIterator, times(1)).emit(
      jsonMatcher("{\"col_str\": \"{'Value':'abc'}\", \"col_int\": 3, \"col_long\": 13}"),
      ArgumentMatchers.eq(JInt.valueOf(0)),
      ArgumentMatchers.eq(JLong.valueOf(0))
    )
    verify(mockedIterator, times(1)).emit(
      jsonMatcher("{\"col_str\": \"{'Value':'hello'}\", \"col_int\": 4, \"col_long\": 14}"),
      ArgumentMatchers.eq(JInt.valueOf(0)),
      ArgumentMatchers.eq(JLong.valueOf(1))
    )
    verify(mockedIterator, times(1)).emit(
      jsonMatcher("{\"col_str\": \"{'Value':'xyz'}\", \"col_int\": 5, \"col_long\": 15}"),
      ArgumentMatchers.eq(JInt.valueOf(0)),
      ArgumentMatchers.eq(JLong.valueOf(2))
    )
  }

  test("run emits records starting from provided offset") {
    val newProperties = properties ++ Map("AS_JSON_DOC" -> "true")
    createCustomTopic(topic)
    publishToKafka(topic, AvroRecord("{'Value':'abc'}", 3, 13))
    publishToKafka(topic, AvroRecord("{'Value':'hello'}", 4, 14))
    publishToKafka(topic, AvroRecord("{'Value':'def'}", 7, 17))
    publishToKafka(topic, AvroRecord("{'Value':'xyz'}", 13, 23))

    // records at 0, 1 are already read, committed
    val mockedIterator = mockExasolIterator(newProperties, Seq(0), Seq(1))
    KafkaTopicDataImporter.run(getMockedMetadata(), mockedIterator)

    verify(mockedIterator, times(2)).emit(any(classOf[Array[Object]]))
    verify(mockedIterator, times(2)).emit(anyString(), anyInt().asInstanceOf[JInt], anyLong().asInstanceOf[JLong])
    verify(mockedIterator, times(1)).emit(
      jsonMatcher("{\"col_str\": \"{'Value':'def'}\", \"col_int\": 7, \"col_long\": 17}"),
      ArgumentMatchers.eq(JInt.valueOf(0)),
      ArgumentMatchers.eq(JLong.valueOf(2))
    )
    verify(mockedIterator, times(1)).emit(
      jsonMatcher("{\"col_str\": \"{'Value':'xyz'}\", \"col_int\": 13, \"col_long\": 23}"),
      ArgumentMatchers.eq(JInt.valueOf(0)),
      ArgumentMatchers.eq(JLong.valueOf(3))
    )
  }

  test("run emits records within min / max records per run") {
    val newProperties = properties ++ Map(
      "MAX_POLL_RECORDS" -> "2",
      "MIN_RECORDS_PER_RUN" -> "2",
      "MAX_RECORDS_PER_RUN" -> "4",
      "AS_JSON_DOC" -> "true"
    )
    createCustomTopic(topic)
    publishToKafka(topic, AvroRecord("{'Value':'abc'}", 3, 13))
    publishToKafka(topic, AvroRecord("{'Value':'hello'}", 4, 14))
    publishToKafka(topic, AvroRecord("{'Value':'def'}", 7, 17))
    publishToKafka(topic, AvroRecord("{'Value':'xyz'}", 13, 23))

    // comsumer in two batches each with 2 records
    val mockedIterator = mockExasolIterator(newProperties, Seq(0), Seq(-1))
    KafkaTopicDataImporter.run(getMockedMetadata(), mockedIterator)

    verify(mockedIterator, times(4)).emit(any(classOf[Array[Object]]))
    verify(mockedIterator, times(4)).emit(anyString(), anyInt().asInstanceOf[JInt], anyLong().asInstanceOf[JLong])
  }

  private[this] def getMockedMetadata(): ExaMetadata = {
    val meta = mock[ExaMetadata]
    when(meta.getOutputColumnCount()).thenReturn(3L)
    when(meta.getOutputColumnType(anyInt())).thenAnswer(new Answer[Class[_]]() {
      override def answer(invocation: InvocationOnMock): Class[_] = {
        val columnIndex = invocation.getArguments()(0).asInstanceOf[JInt]
        Seq(
          classOf[String],
          classOf[JInt],
          classOf[JLong]
        )(columnIndex)
      }
    })
    meta
  }

}

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTopicDataImporterAvroToJsonIT.java
package com.exasol.cloudetl.kafka;

import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import java.util.*;

import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatchers;

class KafkaTopicDataImporterAvroToJsonIT extends KafkaTopicDataImporterAvroIT {
    @Test
    void emitsRecordsFromStartingInitialOffset() throws Exception {
        final Map<String, String> newProperties = withProperty("AS_JSON_DOC", "true");
        createCustomTopic(this.topic);
        publishAvro(this.topic, new AvroRecord("{'Value':'abc'}", 3, 13));
        publishAvro(this.topic, new AvroRecord("{'Value':'hello'}", 4, 14));
        publishAvro(this.topic, new AvroRecord("{'Value':'xyz'}", 5, 15));

        final var iterator = mockExasolIterator(newProperties, List.of(0), List.of(-1L));
        KafkaTopicDataImporter.run(metadata(), iterator);

        verify(iterator, times(3)).emit(any(Object[].class));
        verify(iterator, times(3)).emit(anyString(), anyInt(), anyLong());
        verifyJson(iterator, "{'Value':'abc'}", 3, 13, 0);
        verifyJson(iterator, "{'Value':'hello'}", 4, 14, 1);
        verifyJson(iterator, "{'Value':'xyz'}", 5, 15, 2);
    }

    @Test
    void emitsRecordsStartingFromProvidedOffset() throws Exception {
        final Map<String, String> newProperties = withProperty("AS_JSON_DOC", "true");
        createCustomTopic(this.topic);
        publishAvro(this.topic, new AvroRecord("{'Value':'abc'}", 3, 13));
        publishAvro(this.topic, new AvroRecord("{'Value':'hello'}", 4, 14));
        publishAvro(this.topic, new AvroRecord("{'Value':'def'}", 7, 17));
        publishAvro(this.topic, new AvroRecord("{'Value':'xyz'}", 13, 23));

        final var iterator = mockExasolIterator(newProperties, List.of(0), List.of(1L));
        KafkaTopicDataImporter.run(metadata(), iterator);

        verify(iterator, times(2)).emit(any(Object[].class));
        verify(iterator, times(2)).emit(anyString(), anyInt(), anyLong());
        verifyJson(iterator, "{'Value':'def'}", 7, 17, 2);
        verifyJson(iterator, "{'Value':'xyz'}", 13, 23, 3);
    }

    @Test
    void emitsRecordsWithinMinMaxRecordsPerRun() throws Exception {
        final Map<String, String> newProperties = new LinkedHashMap<>(this.properties);
        newProperties.put("MAX_POLL_RECORDS", "2");
        newProperties.put("MIN_RECORDS_PER_RUN", "2");
        newProperties.put("MAX_RECORDS_PER_RUN", "4");
        newProperties.put("AS_JSON_DOC", "true");
        createCustomTopic(this.topic);
        publishAvro(this.topic, new AvroRecord("{'Value':'abc'}", 3, 13));
        publishAvro(this.topic, new AvroRecord("{'Value':'hello'}", 4, 14));
        publishAvro(this.topic, new AvroRecord("{'Value':'def'}", 7, 17));
        publishAvro(this.topic, new AvroRecord("{'Value':'xyz'}", 13, 23));

        final var iterator = mockExasolIterator(newProperties, List.of(0), List.of(-1L));
        KafkaTopicDataImporter.run(metadata(), iterator);

        verify(iterator, times(4)).emit(any(Object[].class));
        verify(iterator, times(4)).emit(anyString(), anyInt(), anyLong());
    }

    private Map<String, String> withProperty(final String key, final String value) {
        final Map<String, String> result = new LinkedHashMap<>(this.properties);
        result.put(key, value);
        return result;
    }

    private com.exasol.ExaMetadata metadata() throws Exception {
        return KafkaTopicDataImporterJsonToColumnsIT.mockMetadata(String.class, Integer.class, Long.class);
    }

    private void verifyJson(final com.exasol.ExaIterator iterator, final String colStr, final int colInt,
            final long colLong, final long offset) throws Exception {
        verify(iterator).emit(jsonMatcher("{\"col_str\":\"" + colStr + "\",\"col_int\":" + colInt
                + ",\"col_long\":" + colLong + "}"), ArgumentMatchers.eq(Integer.valueOf(0)),
                ArgumentMatchers.eq(Long.valueOf(offset)));
    }
}
